src/serac/numerics/functional/tests/check_gradient.hpp

```diff
@@ -13,100 +13,172 @@
 #include "serac/numerics/functional/functional.hpp"
 
 template <typename T>
-void check_gradient(serac::Functional<T>& f, mfem::Vector& U)
+void check_gradient(serac::Functional<T>& f, mfem::Vector& U, double epsilon = 1.0e-4)
 {
   int seed = 42;
 
   mfem::Vector dU(U.Size());
   dU.Randomize(seed);
 
-  double epsilon = 1.0e-8;
-
-  auto U_plus = U;
-  U_plus.Add(epsilon, dU);
-
-  auto U_minus = U;
-  U_minus.Add(-epsilon, dU);
-
-  mfem::Vector df1 = f(U_plus);
-  df1 -= f(U_minus);
-  df1 /= (2 * epsilon);
-
-  auto [value, dfdU] = f(serac::differentiate_wrt(U));
-  mfem::Vector df2   = dfdU(dU);
+  auto [value, dfdU]   = f(serac::differentiate_wrt(U));
+  mfem::Vector df_jvp1 = dfdU(dU);
```

Comment:
`jvp` and `vjp` are how `jax` describes these kind of operations. I was considering changing our terminology to be more consistent with `jax`, to avoid having a bunch of incompatible terminology for the same ideas.

I'd prefer to not have to annotate `Functional::Gradient::operator()` output at its call site, but I can document that method more thoroughly to clarify that it isn't forming a sparse matrix.